File: services/scanner/src/db/queries.rs
use chrono::Utc;
use uuid::Uuid;

use super::models::{NewObservation, ScanJob, Target};

pub async fn get_scan_job<'e, E>(executor: E, job_id: Uuid) -> sqlx::Result<ScanJob>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query_as::<_, ScanJob>("SELECT * FROM scan_jobs WHERE id = $1")
        .bind(job_id)
        .fetch_one(executor)
        .await
}

pub async fn get_target<'e, E>(executor: E, target_id: Uuid) -> sqlx::Result<Target>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query_as::<_, Target>("SELECT * FROM targets WHERE id = $1")
        .bind(target_id)
        .fetch_one(executor)
        .await
}

pub async fn update_scan_job_status<'e, E>(
    executor: E,
    job_id: Uuid,
    status: &str,
    error_message: Option<&str>,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    let now = Utc::now();
    match status {
        "running" => {
            sqlx::query("UPDATE scan_jobs SET status = $2, started_at = $3 WHERE id = $1")
                .bind(job_id)
                .bind(status)
                .bind(now)
                .execute(executor)
                .await?;
        }
        "completed" | "failed" => {
            sqlx::query(
                "UPDATE scan_jobs SET status = $2, completed_at = $3, error_message = $4 WHERE id = $1",
            )
            .bind(job_id)
            .bind(status)
            .bind(now)
            .bind(error_message)
            .execute(executor)
            .await?;
        }
        _ => {
            sqlx::query("UPDATE scan_jobs SET status = $2 WHERE id = $1")
                .bind(job_id)
                .bind(status)
                .execute(executor)
                .await?;
        }
    }
    Ok(())
}

pub async fn insert_scan<'e, E>(executor: E, target_id: Uuid) -> sqlx::Result<Uuid>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    let row: (Uuid,) =
        sqlx::query_as("INSERT INTO scans (target_id, status) VALUES ($1, 'running') RETURNING id")
            .bind(target_id)
            .fetch_one(executor)
            .await?;
    Ok(row.0)
}

pub async fn complete_scan<'e, E>(executor: E, scan_id: Uuid, sbom_hash: &str) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        "UPDATE scans SET status = 'completed', completed_at = now(), sbom_hash = $2 WHERE id = $1",
    )
    .bind(scan_id)
    .bind(sbom_hash)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn fail_scan<'e, E>(executor: E, scan_id: Uuid, error: &str) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        "UPDATE scans SET status = 'failed', completed_at = now(), sbom_hash = $2 WHERE id = $1",
    )
    .bind(scan_id)
    .bind(error)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn upsert_vulnerability_instance<'e, E>(
    executor: E,
    target_id: Uuid,
    package_name: &str,
    package_version: &str,
    ecosystem: &str,
    advisory_id: &str,
    advisory_source: &str,
) -> sqlx::Result<Uuid>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    let row: (Uuid,) = sqlx::query_as(
        r#"
        INSERT INTO vulnerability_instances
            (target_id, package_name, package_version, ecosystem, advisory_id, advisory_source)
        VALUES ($1, $2, $3, $4, $5, $6)
        ON CONFLICT (target_id, package_name, package_version, ecosystem, advisory_id)
        DO UPDATE SET advisory_source = EXCLUDED.advisory_source
        RETURNING id
        "#,
    )
    .bind(target_id)
    .bind(package_name)
    .bind(package_version)
    .bind(ecosystem)
    .bind(advisory_id)
    .bind(advisory_source)
    .fetch_one(executor)
    .await?;
    Ok(row.0)
}

pub async fn insert_vulnerability_observation<'e, E>(
    executor: E,
    obs: &NewObservation<'_>,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        r#"
        INSERT INTO vulnerability_observations
            (instance_id, scan_id, cvss_score, cvss_vector, epss_score, epss_percentile, decree_score, severity, reachability, is_direct_dep, dep_depth)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
        "#,
    )
    .bind(obs.instance_id)
    .bind(obs.scan_id)
    .bind(obs.cvss_score)
    .bind(obs.cvss_vector)
    .bind(obs.epss_score)
    .bind(obs.epss_percentile)
    .bind(obs.decree_score)
    .bind(obs.severity)
    .bind(obs.reachability)
    .bind(obs.is_direct_dep)
    .bind(obs.dep_depth)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn insert_advisory_fix_version<'e, E>(
    executor: E,
    instance_id: Uuid,
    version: &str,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        "INSERT INTO advisory_fix_versions (instance_id, fixed_version) VALUES ($1, $2)
         ON CONFLICT DO NOTHING",
    )
    .bind(instance_id)
    .bind(version)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn upsert_advisory<'e, E>(
    executor: E,
    advisory_id: &str,
    source: &str,
    raw_json: &serde_json::Value,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        r#"
        INSERT INTO advisories (advisory_id, source, raw_json, fetched_at)
        VALUES ($1, $2, $3, now())
        ON CONFLICT (advisory_id, source) DO UPDATE SET
            raw_json = EXCLUDED.raw_json,
            fetched_at = now()
        "#,
    )
    .bind(advisory_id)
    .bind(source)
    .bind(raw_json)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn insert_advisory_alias<'e, E>(
    executor: E,
    advisory_id: &str,
    alias: &str,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        r#"
        INSERT INTO advisory_aliases (advisory_id, alias)
        VALUES ($1, $2)
        ON CONFLICT (advisory_id, alias) DO NOTHING
        "#,
    )
    .bind(advisory_id)
    .bind(alias)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn insert_dependency_edge<'e, E>(
    executor: E,
    scan_id: Uuid,
    target_id: Uuid,
    from_pkg: &str,
    to_pkg: &str,
    dep_type: &str,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        "INSERT INTO dependency_edges (scan_id, target_id, from_pkg, to_pkg, dep_type)
         VALUES ($1, $2, $3, $4, $5)",
    )
    .bind(scan_id)
    .bind(target_id)
    .bind(from_pkg)
    .bind(to_pkg)
    .bind(dep_type)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn upsert_current_finding_status<'e, E>(
    executor: E,
    instance_id: Uuid,
    target_id: Uuid,
    scan_id: Uuid,
    is_active: bool,
    score: Option<f32>,
    severity: &str,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query(
        r#"
        INSERT INTO current_finding_status
            (instance_id, target_id, latest_scan_id, is_active, last_observed_at, last_score, last_severity, updated_at)
        VALUES ($1, $2, $3, $4, now(), $5, $6, now())
        ON CONFLICT (instance_id) DO UPDATE SET
            latest_scan_id = EXCLUDED.latest_scan_id,
            is_active = EXCLUDED.is_active,
            last_observed_at = EXCLUDED.last_observed_at,
            last_score = EXCLUDED.last_score,
            last_severity = EXCLUDED.last_severity,
            updated_at = now()
        "#,
    )
    .bind(instance_id)
    .bind(target_id)
    .bind(scan_id)
    .bind(is_active)
    .bind(score)
    .bind(severity)
    .execute(executor)
    .await?;
    Ok(())
}

pub async fn insert_outbox_event<'e, E>(
    executor: E,
    stream_name: &str,
    payload: &serde_json::Value,
) -> sqlx::Result<()>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query("INSERT INTO stream_outbox (stream_name, payload) VALUES ($1, $2)")
        .bind(stream_name)
        .bind(payload)
        .execute(executor)
        .await?;
    Ok(())
}

/// Create a new scan_job and return its ID.
pub async fn insert_scan_job<'e, E>(executor: E, target_id: Uuid) -> sqlx::Result<Uuid>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    let row: (Uuid,) = sqlx::query_as("INSERT INTO scan_jobs (target_id) VALUES ($1) RETURNING id")
        .bind(target_id)
        .fetch_one(executor)
        .await?;
    Ok(row.0)
}

/// Fetch the latest EPSS score and percentile for a CVE ID.
pub async fn get_latest_epss<'e, E>(executor: E, cve_id: &str) -> sqlx::Result<Option<(f32, f32)>>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    sqlx::query_as(
        "SELECT epss_score, epss_percentile FROM advisory_epss_snapshots WHERE cve_id = $1 ORDER BY epss_date DESC LIMIT 1",
    )
    .bind(cve_id)
    .fetch_optional(executor)
    .await
}

File: services/oracle/internal/db/queries.go
package db

import (
	"context"
	"encoding/json"
	"log/slog"
	"time"

	"github.com/google/uuid"
)

// ListTargets returns all targets from the database.
func (d *DB) ListTargets(ctx context.Context) ([]Target, error) {
	rows, err := d.Pool.Query(ctx, `
		SELECT id, project_id, name, target_type, source_ref, branch, subpath, exposure_class
		FROM targets
		ORDER BY name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var targets []Target
	for rows.Next() {
		var t Target
		if err := rows.Scan(&t.ID, &t.ProjectID, &t.Name, &t.TargetType,
			&t.SourceRef, &t.Branch, &t.Subpath, &t.ExposureClass); err != nil {
			return nil, err
		}
		targets = append(targets, t)
	}
	return targets, rows.Err()
}

// UpsertTarget inserts or updates a target, returning its ID.
func (d *DB) UpsertTarget(ctx context.Context, projectID uuid.UUID, name, targetType string, sourceRef, branch *string) (uuid.UUID, error) {
	var id uuid.UUID
	err := d.Pool.QueryRow(ctx, `
		INSERT INTO targets (project_id, name, target_type, source_ref, branch)
		VALUES ($1, $2, $3, $4, $5)
		ON CONFLICT (project_id, name) DO UPDATE SET
			target_type = EXCLUDED.target_type,
			source_ref  = EXCLUDED.source_ref,
			branch      = EXCLUDED.branch
		RETURNING id`, projectID, name, targetType, sourceRef, branch).Scan(&id)

	if err != nil {
		// If ON CONFLICT DO NOTHING returned no rows, look up existing
		err2 := d.Pool.QueryRow(ctx, `
			SELECT id FROM targets WHERE project_id = $1 AND name = $2`,
			projectID, name).Scan(&id)
		if err2 != nil {
			return uuid.Nil, err2
		}
	}
	return id, nil
}

// EnsureProject creates a project if it doesn't exist, returning its ID.
func (d *DB) EnsureProject(ctx context.Context, name string) (uuid.UUID, error) {
	var id uuid.UUID
	err := d.Pool.QueryRow(ctx, `
		INSERT INTO projects (name)
		VALUES ($1)
		ON CONFLICT (name) DO NOTHING
		RETURNING id`, name).Scan(&id)

	if err != nil {
		err2 := d.Pool.QueryRow(ctx, `SELECT id FROM projects WHERE name = $1`, name).Scan(&id)
		if err2 != nil {
			return uuid.Nil, err2
		}
	}
	return id, nil
}

// AcquireLease attempts to acquire an exclusive lease for a target.
// Returns true if acquired, false if already held by another.
func (d *DB) AcquireLease(ctx context.Context, targetID uuid.UUID, holderID string, ttl time.Duration) (bool, error) {
	tag, err := d.Pool.Exec(ctx, `
		INSERT INTO job_leases (target_id, holder_id, acquired_at, expires_at)
		VALUES ($1, $2, now(), now() + $3::interval)
		ON CONFLICT (target_id) DO UPDATE
		SET holder_id = EXCLUDED.holder_id,
		    acquired_at = EXCLUDED.acquired_at,
		    expires_at = EXCLUDED.expires_at
		WHERE job_leases.expires_at < now()
		   OR job_leases.holder_id = EXCLUDED.holder_id`,
		targetID, holderID, ttl.String())
	if err != nil {
		return false, err
	}
	return tag.RowsAffected() > 0, nil
}

// ReleaseLease releases a lease held by the given holder.
func (d *DB) ReleaseLease(ctx context.Context, targetID uuid.UUID, holderID string) error {
	_, err := d.Pool.Exec(ctx, `
		DELETE FROM job_leases
		WHERE target_id = $1 AND holder_id = $2`,
		targetID, holderID)
	return err
}

// ClearExpiredLeases removes all leases whose TTL has passed.
// Called on startup to reclaim leases left by a previous oracle instance.
func (d *DB) ClearExpiredLeases(ctx context.Context) error {
	tag, err := d.Pool.Exec(ctx, `DELETE FROM job_leases WHERE expires_at < now()`)
	if err != nil {
		return err
	}
	if tag.RowsAffected() > 0 {
		slog.Info("cleared expired leases", "count", tag.RowsAffected())
	}
	return nil
}

// UpdateLeaseJobID sets the job_id on a lease.
func (d *DB) UpdateLeaseJobID(ctx context.Context, targetID uuid.UUID, jobID uuid.UUID) error {
	_, err := d.Pool.Exec(ctx, `
		UPDATE job_leases SET job_id = $1 WHERE target_id = $2`,
		jobID, targetID)
	return err
}

// GetCurrentObservations returns observations for a given scan.
func (d *DB) GetCurrentObservations(ctx context.Context, scanID uuid.UUID) ([]Observation, error) {
	rows, err := d.Pool.Query(ctx, `
		SELECT vo.instance_id, vi.package_name, vi.package_version, vi.ecosystem,
		       vi.advisory_id, vo.cvss_score, vo.epss_score, vo.decree_score,
		       vo.severity, vo.is_direct_dep, vo.dep_depth
		FROM vulnerability_observations vo
		JOIN vulnerability_instances vi ON vi.id = vo.instance_id
		WHERE vo.scan_id = $1`, scanID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var obs []Observation
	for rows.Next() {
		var o Observation
		if err := rows.Scan(&o.InstanceID, &o.PackageName, &o.PackageVersion,
			&o.Ecosystem, &o.AdvisoryID, &o.CVSSScore, &o.EPSSScore,
			&o.DecreeScore, &o.Severity, &o.IsDirectDep, &o.DepDepth); err != nil {
			return nil, err
		}
		obs = append(obs, o)
	}
	return obs, rows.Err()
}

// GetPreviousCompletedScanID returns the scan_id of the most recent completed scan
// for the target before the given scan.
func (d *DB) GetPreviousCompletedScanID(ctx context.Context, targetID, currentScanID uuid.UUID) (uuid.UUID, error) {
	var prevID uuid.UUID
	err := d.Pool.QueryRow(ctx, `
		SELECT id FROM scans
		WHERE target_id = $1
		  AND status = 'completed'
		  AND id != $2
		ORDER BY completed_at DESC
		LIMIT 1`, targetID, currentScanID).Scan(&prevID)
	return prevID, err
}

// GetExploitLinkedCVEs returns a set of CVE IDs that have exploit links.
func (d *DB) GetExploitLinkedCVEs(ctx context.Context, cveIDs []string) (map[string]bool, error) {
	if len(cveIDs) == 0 {
		return map[string]bool{}, nil
	}
	rows, err := d.Pool.Query(ctx, `
		SELECT DISTINCT cve_id FROM exploit_cve_links
		WHERE cve_id = ANY($1)`, cveIDs)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	result := make(map[string]bool)
	for rows.Next() {
		var cveID string
		if err := rows.Scan(&cveID); err != nil {
			return nil, err
		}
		result[cveID] = true
	}
	return result, rows.Err()
}

// InsertDisappearance records a vulnerability disappearance.
func (d *DB) InsertDisappearance(ctx context.Context, instanceID, scanID uuid.UUID) error {
	_, err := d.Pool.Exec(ctx, `
		INSERT INTO vulnerability_disappearances (instance_id, scan_id)
		VALUES ($1, $2)`, instanceID, scanID)
	return err
}

// InsertOutboxEvent inserts an event into stream_outbox for publishing to Redis.
func (d *DB) InsertOutboxEvent(ctx context.Context, streamName string, payload any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	_, err = d.Pool.Exec(ctx, `
		INSERT INTO stream_outbox (stream_name, payload)
		VALUES ($1, $2)`, streamName, data)
	return err
}

// InsertDeliveryLog inserts a notification delivery log entry.
func (d *DB) InsertDeliveryLog(ctx context.Context, rec DeliveryRecord) error {
	_, err := d.Pool.Exec(ctx, `
		INSERT INTO notification_delivery_log
			(target_id, advisory_id, diff_kind, channel, status, attempts, dedup_key)
		VALUES ($1, $2, $3, $4, $5, $6, $7)`,
		rec.TargetID, rec.AdvisoryID, rec.DiffKind, rec.Channel, rec.Status, rec.Attempts, rec.DedupKey)
	return err
}

// CheckDedup returns true if a delivery with this dedup_key+channel already succeeded.
func (d *DB) CheckDedup(ctx context.Context, dedupKey, channel string) (bool, error) {
	var exists bool
	err := d.Pool.QueryRow(ctx, `
		SELECT EXISTS(
			SELECT 1 FROM notification_delivery_log
			WHERE dedup_key = $1 AND channel = $2 AND status = 'delivered'
		)`, dedupKey, channel).Scan(&exists)
	return exists, err
}

// UpdateDeliveryStatus updates the status of a delivery log entry.
func (d *DB) UpdateDeliveryStatus(ctx context.Context, id uuid.UUID, status string) error {
	now := time.Now()
	_, err := d.Pool.Exec(ctx, `
		UPDATE notification_delivery_log
		SET status = $1, last_attempt_at = $2, attempts = attempts + 1,
		    delivered_at = CASE WHEN $1 = 'delivered' THEN $2 ELSE delivered_at END
		WHERE id = $3`, status, now, id)
	return err
}

// GetTargetName returns the name of a target by ID.
func (d *DB) GetTargetName(ctx context.Context, targetID uuid.UUID) (string, error) {
	var name string
	err := d.Pool.QueryRow(ctx, `SELECT name FROM targets WHERE id = $1`, targetID).Scan(&name)
	return name, err
}

// GetTargetProjectID returns the project ID for a target.
func (d *DB) GetTargetProjectID(ctx context.Context, targetID uuid.UUID) (uuid.UUID, error) {
	var projectID uuid.UUID
	err := d.Pool.QueryRow(ctx, `SELECT project_id FROM targets WHERE id = $1`, targetID).Scan(&projectID)
	return projectID, err
}

// GetFixVersions returns fix versions for a vulnerability instance.
func (d *DB) GetFixVersions(ctx context.Context, instanceID uuid.UUID) ([]string, error) {
	rows, err := d.Pool.Query(ctx, `
		SELECT fixed_version FROM advisory_fix_versions WHERE instance_id = $1`,
		instanceID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var versions []string
	for rows.Next() {
		var v string
		if err := rows.Scan(&v); err != nil {
			return nil, err
		}
		versions = append(versions, v)
	}
	return versions, rows.Err()
}
